### ISSUE TITLE
Regression: node halts with fatal error when importing snapshot (recoverable)

### ISSUE THREAD

**Issue by jsign** [filecoin-project/lotus#4850]
While updating a bunch of Lotus nodes to v1.1.3, I had to also import some chain snapshots. I feel quite confident of never seeing this in <=1.1.2.

Looks like I get the following situation consistently:
1. `lotus daemon --import-snapshot ...`
2. Snapshot importing goes fine.
3. Metadata cache clearing goes fine
4. When its about to continue starting the node:
```
ERROR: initializing node: starting node: could not build arguments for function "github.com/filecoin-project/lotus/node/modules/lp2p".StartListening.func1 (/lotus/node/modules/lp2p/addrs.go:98): failed to build host.Host: could not build arguments for function "reflect".makeFuncStu
b (/usr/local/go/src/reflect/asm_amd64.s:12): failed to build lp2p.BaseIpfsRouting: could not build arguments for function "reflect".makeFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to build dtypes.NetworkName: could not build arguments for function "reflect".makeFun
cStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to build *store.ChainStore: could not build arguments for function "reflect".makeFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to build dtypes.ChainBlockstore: could not build arguments for function "reflect".ma
keFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to build dtypes.ChainRawBlockstore: received non-nil error from function "reflect".makeFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to open badger blockstore: Cannot acquire directory lock on "/data/.lotus
/datastore/chain".  Another process is using this Badger database.: resource temporarily unavailable
```
I can't imagine any other process messing with Badger or similar... mostly because this happened to more than 4 Lotus nodes in different VMs.

After the node stops, if you `lotus daemon` (without the snapshot importing flag), the node starts again without problems and all looks good.

In summary: after importing the snapshot the node halts. You can start again and things are normal.

**Comment by raulk** [comment]
Thanks for pointing out the regression! Luckily you can start the node afterwards, so this is non-blocking as you duly point out. I'll take a look soon.

**Comment by raulk** [comment]
Context: we shipped a batch of blockstore improvements, and moved to a native badger blockstore implementation for added mechanical sympathy. I see that `ImportChain` _does_ close the blockstore properly, so I'm not sure what's going on here. 

To help me narrow this down, did you happen to see a message similar to: `failed to close blockstore` in your log output?

**Comment by jennijuju** [comment]
@hannahhoward could you please check too?

**Comment by jsign** [comment]
> Context: we shipped a batch of blockstore improvements, and moved to a native badger blockstore implementation for added mechanical sympathy.
> 
> I see that `ImportChain` _does_ close the blockstore properly, so I'm not sure what's going on here.
> 
> To help me narrow this down, did you happen to see a message similar to: `failed to close blockstore` in your log output?

Uh, I don't have history of log lines in these cases since these imports weren't _teeed_ or similar. I can run some import locally in ~1.5hr and see if can reproduce locally. I'll let you know!

**Comment by jsign** [comment]
@raulk ,
```
 10.80 GiB / 12.21 GiB [===========================================================================================================================================================================================================>--------------------------]  88.42% 92.34 MiB/s 00m15s2020-11-13T17:41:12.504-0300    INFO    badgerbs        v2@v2.2007.2/db.go:1030 Storing value log head: {Fid:14 Len:33 Offset:549952285}

 11.23 GiB / 12.21 GiB [===================================================================================================================================================================================================================>------------------]  91.98% 92.34 MiB/s 00m10s2020-11-13T17:41:17.424-0300    INFO    badgerbs        v2@v2.2007.2/db.go:1030 Storing value log head: {Fid:15 Len:33 Offset:65623498}

 11.79 GiB / 12.21 GiB [=============================================================================================================================================================================================================================>--------]  96.51% 92.85 MiB/s 00m04s2020-11-13T17:41:22.855-0300    INFO    badgerbs        v2@v2.2007.2/db.go:1030 Storing value log head: {Fid:15 Len:33 Offset:688945757}

 11.83 GiB / 12.21 GiB [==============================================================================================================================================================================================================================>-------]  96.86% 92.90 MiB/s 00m04s2020-11-13T17:41:23.143-0300    INFO    badgerbs        v2@v2.2007.2/levels.go:1000     [Compactor: 0] Running compaction: {level:0 score:1 dropPrefixes:[]} for level: 0

 11.99 GiB / 12.21 GiB [=================================================================================================================================================================================================================================>----]  98.15% 92.85 MiB/s 00m02s2020-11-13T17:41:25.074-0300    INFO    badgerbs        v2@v2.2007.2/levels.go:962      LOG Compact 0->1, del 8 tables, add 7 tables, took 1.930705796s

2020-11-13T17:41:25.074-0300    INFO    badgerbs        v2@v2.2007.2/levels.go:1010     [Compactor: 0] Compaction for level: 0 DONE
2020-11-13T17:41:25.074-0300    INFO    badgerbs        v2@v2.2007.2/levels.go:1000     [Compactor: 0] Running compaction: {level:1 score:1.7069128081202507 dropPrefixes:[]} for level: 1

 12.08 GiB / 12.21 GiB [===================================================================================================================================================================================================================================>--]  98.91% 92.73 MiB/s 00m01s2020-11-13T17:41:26.165-0300    INFO    badgerbs        v2@v2.2007.2/levels.go:962      LOG Compact 1->2, del 4 tables, add 4 tables, took 1.090904479s

2020-11-13T17:41:26.165-0300    INFO    badgerbs        v2@v2.2007.2/levels.go:1010     [Compactor: 0] Compaction for level: 1 DONE
2020-11-13T17:41:26.165-0300    INFO    badgerbs        v2@v2.2007.2/levels.go:1000     [Compactor: 0] Running compaction: {level:1 score:1.4466810040175915 dropPrefixes:[]} for level: 1

 12.13 GiB / 12.21 GiB [===========================================================================================================================================================================================================================================>-]  99.35% 92.72 MiB/s2020-11-13T17:41:26.732-0300    INFO    badgerbs        v2@v2.2007.2/levels.go:962      LOG Compact 1->2, del 2 tables, add 2 tables, took 566.653353ms

2020-11-13T17:41:26.732-0300    INFO    badgerbs        v2@v2.2007.2/levels.go:1010     [Compactor: 0] Compaction for level: 1 DONE
2020-11-13T17:41:26.732-0300    INFO    badgerbs        v2@v2.2007.2/levels.go:1000     [Compactor: 0] Running compaction: {level:1 score:1.186449807137251 dropPrefixes:[]} for level: 1

 12.20 GiB / 12.21 GiB [=============================================================================================================================================================================================================================================]  99.90% 92.68 MiB/s2020-11-13T17:41:27.531-0300    INFO    badgerbs        v2@v2.2007.2/levels.go:962      LOG Compact 1->2, del 3 tables, add 3 tables, took 799.266978ms

2020-11-13T17:41:27.531-0300    INFO    badgerbs        v2@v2.2007.2/levels.go:1010     [Compactor: 0] Compaction for level: 1 DONE
 12.21 GiB / 12.21 GiB [=======================================================================================================================================================================================================================================] 100.00% 92.68 MiB/s 2m14s
2020-11-13T17:41:27.622-0300    INFO    chainstore      store/store.go:486      clearing block validation cache...
2020-11-13T17:41:27.630-0300    INFO    chainstore      store/store.go:525      3102 block validation entries cleared.
2020-11-13T17:41:45.537-0300    WARN    chainstore      store/store.go:467      no heaviest tipset found, using [bafy2bzacecnamqgqmifpluoeldx7zzglxcljo6oja4vrmtj7432rphldpdmm2]
2020-11-13T17:41:45.537-0300    INFO    chainstore      store/store.go:472      New heaviest tipset! [bafy2bzacecnamqgqmifpluoeldx7zzglxcljo6oja4vrmtj7432rphldpdmm2] (height=0)
2020-11-13T17:41:45.538-0300    INFO    main    lotus/daemon.go:474     accepting [bafy2bzaceayd56jdy3hgo4kgrjykrwqlmunojql23gf2rxrngilgwqstom2pq bafy2bzaced6csjmhb5y4i6xhunkkmrp7ofdjf5qwce27kjg4bnbjlpkgzp4pk bafy2bzacedscooumji6ybpc6ngscxrfxv6lzypb5edovuagulnw35djhnioka bafy2bzaceasroxqnphxrx3g7g4t4psjqqvqegywmpxgwxh72lts62o6225x7u bafy2bzacea4a2x63wgckuenw76tnbm5hevoyf5mepohkixedjf52p34t5v36w bafy2bzacecvsws32ouxhcowjuneb2l2xciccodp6sux2pqvzkg6e2cwdxerdu] as new head
2020-11-13T17:41:45.538-0300    INFO    chainstore      store/store.go:472      New heaviest tipset! [bafy2bzaceayd56jdy3hgo4kgrjykrwqlmunojql23gf2rxrngilgwqstom2pq bafy2bzaced6csjmhb5y4i6xhunkkmrp7ofdjf5qwce27kjg4bnbjlpkgzp4pk bafy2bzacedscooumji6ybpc6ngscxrfxv6lzypb5edovuagulnw35djhnioka bafy2bzaceasroxqnphxrx3g7g4t4psjqqvqegywmpxgwxh72lts62o6225x7u bafy2bzacea4a2x63wgckuenw76tnbm5hevoyf5mepohkixedjf52p34t5v36w bafy2bzacecvsws32ouxhcowjuneb2l2xciccodp6sux2pqvzkg6e2cwdxerdu] (height=232781)
2020-11-13T17:41:45.568-0300    INFO    badger  v2@v2.2007.2/levels.go:183      All 0 tables opened in 0s

2020-11-13T17:41:45.570-0300    INFO    badger  v2@v2.2007.2/value.go:1158      Replaying file id: 0 at offset: 0

2020-11-13T17:41:45.570-0300    INFO    badger  v2@v2.2007.2/value.go:1178      Replay took: 4.278µs

2020-11-13T17:41:45.639-0300    INFO    badger  v2@v2.2007.2/levels.go:183      All 0 tables opened in 0s

2020-11-13T17:41:45.640-0300    INFO    badger  v2@v2.2007.2/value.go:1158      Replaying file id: 0 at offset: 0

2020-11-13T17:41:45.640-0300    INFO    badger  v2@v2.2007.2/value.go:1178      Replay took: 1.483µs

ERROR: initializing node: starting node: could not build arguments for function "github.com/filecoin-project/lotus/node/modules/lp2p".StartListening.func1 (/home/ignacio/code/lotus/node/modules/lp2p/addrs.go:98): failed to build host.Host: could not build arguments for function "reflect".makeFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to build lp2p.BaseIpfsRouting: could not build arguments for function "reflect".makeFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to build dtypes.NetworkName: could not build arguments for function "reflect".makeFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to build *store.ChainStore: could not build arguments for function "reflect".makeFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to build dtypes.ChainBlockstore: could not build arguments for function "reflect".makeFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to build dtypes.ChainRawBlockstore: received non-nil error from function "reflect".makeFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to open badger blockstore: Cannot acquire directory lock on "/home/ignacio/.lotus/datastore/chain".  Another process is using this Badger database.: resource temporarily unavailable
```
Logs from above are the usual importing output...

**Comment by schomatis** [comment]
https://github.com/filecoin-project/lotus/issues/4428 (there might be others like it)

**Comment by raulk** [comment]
Found the root cause. We close the store after the import, but the next call to `repo.Blockstore()` will not open it again, because the opening is enclosed in a sync.Once.

**Comment by raulk** [comment]
This should be fixed by #4872, in case you wanna try it out, @jsign. Thanks for the report!

**Comment by jsign** [comment]
@raulk, I verified that in v1.2.0 this was solved. 👍🏼

**Comment by ohJohanZ** [comment]
Failed again in v1.2.2 and v1.2.3

**Comment by chenhonghe** [comment]
lotus version 1.5.2+mainnet+git.09e78a0fe

`ERROR: initializing node: starting node: could not build arguments for function "github.com/filecoin-project/lotus/node/modules/lp2p".StartListening.func1 (/home/pool/lotus/node/modules/lp2p/addrs.go:98): failed to build host.Host: could not build arguments for function "reflect".makeFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to build lp2p.BaseIpfsRouting: could not build arguments for function "reflect".makeFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to build dtypes.NetworkName: could not build arguments for function "reflect".makeFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to build *store.ChainStore: could not build arguments for function "reflect".makeFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to build dtypes.ChainBlockstore: could not build arguments for function "reflect".makeFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to build dtypes.UniversalBlockstore: received non-nil error from function "reflect".makeFuncStub (/usr/local/go/src/reflect/asm_amd64.s:12): failed to open badger blockstore: file does not exist for table 728018`

**How do I start lotus???**
